Export: better performance by avoiding pointless image->python data->image roundtrips

scurest opened the pull request:
This PR creates a "happy path" for exporting images where we don't have to round-trip through Python data, which is much faster and uses much less memory.

Fixes #790 and helps with export times generally.

## Problem

The glTF spec requires that eg. the metalllicRoughness texture have the metalness in B and the roughness in G. The exporter will assemble the appropriate image for this even if the material graph reads the metalness from, say, the R channel of one image and the roughness from the G channel of another.

To do this, *all* images are decoded to python data. The code sequence for this looks roughly like this:

    creating export image:
        pixel_data = np.array(blender_image.pixels[:])              (★1)

    pixel data may be modified to insert channels

    encoding export image:
        if created from blender image that's on disk:
            read back data from the disk

        tmp_image = bpy.data.images.new(...)
        tmp_image.pixels = pixel_data.img.flatten().tolist()        (★2)
        save tmp_image to disk and read it back

The lines marked with a ★ are slow and eat up loooots of memory. Eg. `blender_image.pixels[:]` is going to allocate a tuple of width×height×channels floats. It's really bad.

As you can see, sometimes it was possible to skip ★2 but everything did ★1.

However, for many images, like all color textures, normals, or metal/rough/occlusion textures that are already packed the way the exporter likes, this is totally wasted effort: the image data is converted to python data, nothing gets changed, and the python data is converted back to a new temp image, which is then encoded to disk and read back.

In this case it would have been better to just encode the original image!

## Solution

Change `ExportImage` to store some declarative data saying how it's channels should be filled in (ex. it says its R channel should be filled by the G channel of such-and-such Blender image) instead of storing a numpy array. I tried to write a good docstring here, so see it for details.

Then when it comes time to encode, we examine that data and determine whether all we need to do is save some existing Blender image (happy path), or whether we need to do the pixel manipulation to repack the channels (unhappy path).

The unhappy path is basically the same as before, but the happy path can be significantly better :)

## Results

Results seem pretty good. Obviously they depend on how much you can stay on the happy path but since that should include the most common texture types (base color, normal, etc.) it should be pretty often.

Export times:

| | Before | After|
|:---|---|---|
| #790 (8192x8192 normalmap) | OOM | 3s |
| DamagedHelmet | 15s | 7s |
| VC | 17s | 6s |
| Sponza | 91s | 38s |

Reviews and testing are appreciated!

julienduroure commented:
I will have a look, but because material/texture is not my main domain of expertise, I hope @emackey or @donmccurdy can have a look

donmccurdy commented on addons/io_scene_gltf2/blender/exp/gltf2_blender_image.py:
```diff
@@ -17,118 +17,195 @@
 
-        self._img = np.concatenate([self.img, other.img], axis=2)
+    def fill_white(self, chan: Channel):
+        self.fills[chan] = White()
```
This is minor, but the naming of "attach" and "Attachment" don't feel very clear to me. On the other hand, `fill_white(Channel.B)` is quite easy to understand. Since the Attachment class is only used once outside this class, maybe replace it with a similar function that accepts an image?

```
image.fill_image(image, Channel.R)
image.fill_white(Channel.G)
```

If the container class is useful regardless, maybe names like ImageFill or SourceImage would be more obvious?

scurest replied:
I'd like to improve the names too. `Attachment` was loosely based on color attachments for a render pass.

How about this then?

````py
# Outside this file
export_image.fill_image(image, dst_chan=dst_chan, src_chan=src_chan)
export_image.fill_white(dst_chan=dst_chan)

# Only used inside this file
class FillImage # fills with an image channel
class FillWhite # fills with all 1.0s
````

donmccurdy replied:
Sounds good to me!

scurest replied:
Alright, done. I also renamed `has_channel`->`is_filled`.

scurest commented:
I wrote a [bash script](https://gist.github.com/scurest/c10a60d362a3b9fe52eef0d55eea9471) for testing if anyone else wants to try. Make sure you have the faster-export-image branch, cd in the glTF-Blender-IO folder, and run `test.sh FILE`. FILE will be loaded and then exported with both master and the faster-export-image branch, and the .pngs will be compared with imagemagick.

So far the only difference I have found is desirable IMO: this branch will sometimes figure out it can use an existing Blender image without change, where master would have changed it.

For example, if you import MetalRoughSpheres.glb, you get this image

![blender](https://user-images.githubusercontent.com/11024420/70076147-c704ad80-15c3-11ea-87b8-525973fd3327.png)

On master, if you export, the red channel will have been filled with ones

![out](https://user-images.githubusercontent.com/11024420/70076243-fadfd300-15c3-11ea-907c-3555cc9539e2.png)

whereas on this branch it exports exactly as it was.

scurest commented:
Oh, here's a difference: when Blender textures are missing, it was possible to have an empty ExportImage which would export as a single pixel image. In master, the image is just thrown away in this case. I'll add a commit to fix it.

scurest commented:
So I tested it with dirty images (ie. images that have been painted on and not saved; in the Image Editor there will be a `*` next to the Image menu if the current image is dirty) and I found another bug. After export, any image that goes down the happy path will have its dirty flag cleared since it was saved to a temp file.

This can be fixed by just doing `image.pixels[0] = image.pixels[0]` on images that were dirty to set their dirty flags again. This is can be pretty slow for some reason. Luckily we only pay this cost if an image was dirty. I added a commit to do this.

Btw, you might have noticed that the first commit also fixed a bug in master where it would use an image from disk even if the image was dirty.

scurest commented:
Saving a copy of an image turns out to be surprisingly hard :(

Besides clearing the dirty flag, saving images would also turn generated images into from-files-on-disk images, so I've given up on using `image.save()` to save images.

Instead I'm now using `image.save_render()` which doesn't modify the image at all so it won't have any of those problems. It does use the render settings from a scene though, so this now creates a temp scene, sets all the settings we need, and then uses `save_render()`. See the commit message on 685e9ec for details.

(Doing it this way is slightly slower unfortunately; Sponza has gone to about 40s vs 38s before; the others I tested are about the same.) 

The reason the tests are failing on 2.79 is that the test expects an RGB image and somewhere between 2.79 and 2.8 Blender learned to do

    if you call img.save_render()
        if the render color mode is RGBA
            if img is an RGB image
                save the image as an RGB file instead of an RGBA file

which makes 2.8 pass, but since 2.79 doesn't know how to do this, the PNG files it saves when it saves a copy will be RGBA images instead of RGB. As far as I could tell, there is no field that exposes if the image data is RGB or RGBA on a Blender image, so I don't know how I could fix this. (Other than having an unnecessary A channel though, I don't think there's anything actually wrong with 2.79).

I'd really love any help writing `_save_copy_of_image`. Does anyone else have any ideas? Is there some way to ask what the correct way to save a copy of an image from Python is?

julienduroure commented:
>  saving images would also turn generated images into from-files-on-disk images

You can get back to generative using bpy.data.images['Untitled'].source = 'GENERATED'

> clearing the dirty flag

This is dirty too, but we can change pixels[0] to any value and back to original one, it will get back the  dirty flag

scurest commented:
> You can get back to generative using bpy.data.images['Untitled'].source = 'GENERATED'

You can't if the image was dirty though. When you set `image.source = 'GENERATED'` the image gets recreated from the generation parameters, which means any changes you painted on top of it would be lost.

> This is dirty too, but we can change pixels[0] to any value and back to original one, it will get back the dirty flag

Yeah, just `image.pixels[0] = image.pixels[0]` works (though like I said before, it is pretty slow; seems anything touching `pixels` is).

The other problem is I don't know if this is an exhaustive list of things that are messed up by `image.save()`.

emackey commented:
@scurest Is `image.save_render()` only blocked by the 2.79 behavior described in your [comment](https://github.com/KhronosGroup/glTF-Blender-IO/pull/798#issuecomment-562442020) above?  We'll be dropping 2.79 soon, as you know.

The other thing I thought of is that the UI actually has something similar to what we want, saving an image as a copy from a menu item.  There's a matching Python operator:

> bpy.ops.image.save_as(save_as_render=False, copy=True, filepath="/path/file.png", relative_path=True, show_multiview=False, use_multiview=False)

I know that addons are discouraged from using `ops` for stuff like this, but it means there's code in there somewhere ([here](https://developer.blender.org/diffusion/B/browse/master/source/blender/editors/space_image/image_ops.c$1919)? [here](https://developer.blender.org/diffusion/B/browse/master/source/blender/blenkernel/intern/image_save.c$161)?) for this, even if there doesn't seem to be a non-ops Python binding for it.

Let's not let version 2.79 hold us back if that's the only stumbling block on `save_render` though.

scurest commented:
>  There's a matching Python operator

Yeah, I'd love to use it but I can't figure out how. Do you know?

> Is image.save_render() only blocked by the 2.79 behavior described in your comment above?

AFAIK yes.

I'm not sure if the 2.79 behaviour should really be considered blocking. It's unfortunate that an alpha channel is added when it isn't needed, but it doesn't really hurt anything except the file size. This also only happens in what should be the uncommon case, when we can't find an existing file to use for the image (eg. it was packed from channels, it was dirty, it was generated).

If another way can't be found... If the test could be relaxed to only check if the images are pixel-for-pixel equal instead of checking if the files are byte-for-byte equal, it should pass on 2.79 too. (I think that would be a good idea anyway; checking if the PNGs files are byte-for-byte equal is too restrictive. For example, the tests will break even if all we do is change the compression setting for PNGs since the file will get encoded differently even though obviously the pixels don't change.)

julienduroure commented:
> Yeah, I'd love to use it but I can't figure out how. Do you know?

You have to override the context, see https://docs.blender.org/api/current/bpy.ops.html#overriding-context

emackey commented:
> > There's a matching Python operator
> 
> Yeah, I'd love to use it but I can't figure out how. Do you know?

I haven't used these much but my understanding is that `bpy.ops.*` is tied specifically to user UI operations, such as saving an image from an Image Editor or UV Editor UI panel.  This may be a dead-end.

> checking if the PNGs files are byte-for-byte equal is too restrictive

Yeah, I put in the byte-for-byte check when I was looking to test a "happy path" where the file was simply copied from the source, and then I probably reused it and never wrote a more detailed pixel-for-pixel check.  Some sort of relaxing of tests would be fine with me, we could even isolate this test to only Blender 2.8x and above, and skip it for 2.79.

scurest commented:
@julienduroure Can you post the code for it?

julienduroure commented:
@scurest Have a look here: 
https://gist.github.com/julienduroure/83661b26fa5e1e442dcdd7873936a5a6

I didn't test in the context of addon exporting

scurest commented:
@julienduroure Thanks a ton! Do you know how to control the file format?

scurest commented:
Actually, you know what, sorry, I've dragged this PR off into the weeds. I didn't expect saving a copy of an image to be so hairy, but this case is not that important. We can just do

    tmp_img = image.copy()
    tmp_img.filepath/file_format = whatever
    tmp_img.save()

which still avoids going through Python data like master has to.

The important thing here is the restructuring of ExportImage to hold a description of how to build the image instead of holding raw pixel data so that we can avoid decoding every single image to a Python tuple. That's plenty for one PR.

I'm going to close this and open a fresh PR with just that. Trying to figure out how to save a copy of an image can always be done separately. Sorry for the noise.
